genocrate/commands/validate.py

```diff
@@ -0,0 +1,159 @@
+import json
+import os
+import sys
+import hashlib
+import multiprocessing
+
+import click
+from genocrate.main import cli
+import bagit
+
+# Block size used when reading files for hashing
+HASH_BLOCK_SIZE = 512 * 1024
+
+
+def is_valid_bagit(path: str, processes: int = 1) -> bool:
+    """
+    Validate all bags in a directory using bagit-python.
+
+    :param path: Path to the directory containing bags
+    :param processes: Number of parallel processes to use
+    :return: True if valid, False otherwise
+    """
+    bag = bagit.Bag(path)
+    return bag.is_valid(processes=processes)
+
+
+def _calc_hashes(filename: str) -> tuple:
+    """
+    Calculate the MD5 hash of a file.
+
+    :param filename: Path to the file
+    :return: Tuple of (filename, md5 hash)
+    """
+    md5 = hashlib.md5()
+    with open(filename, 'rb') as f:
+        while chunk := f.read(HASH_BLOCK_SIZE):
+            md5.update(chunk)
+    return filename, md5.hexdigest()
+
+
+def read_manifest(manifest_path: str) -> dict:
+    """
+    Read a manifest file and return a dict mapping filenames to checksums.
+
+    :param manifest_path: Path to the manifest file
+    :return: Dict of {filename: checksum}
+    """
+    expected = {}
+    with open(manifest_path, 'r') as f:
+        for line in f:
+            line = line.strip()
+            if not line or line.startswith('#'):
+                continue
+            parts = line.split()
+            if len(parts) < 2:
+                continue
+            checksum, filename = parts[0], parts[-1]
+            expected[filename] = checksum
+    return expected
+
+
+def worker_init():
+    """Initializer for multiprocessing workers (placeholder)."""
+    pass
+
+
+def compute_hashes(file_list, processes: int = 1) -> list:
+    """
+    Compute hashes for a list of files, optionally in parallel.
+
+    :param file_list: List of file paths
+    :param processes: Number of parallel processes to use
+    :return: List of (filename, hash) tuples
+    """
+    if processes == 1:
+        return [_calc_hashes(f) for f in file_list]
+    with multiprocessing.Pool(processes, initializer=worker_init) as pool:
+        return pool.map(_calc_hashes, file_list)
+
+
+def list_files_exclude_manifest(manifest_path: str) -> list:
+    """
+    List all files in the manifest's directory, excluding the manifest itself.
+
+    :param manifest_path: Path to the manifest file
+    :return: List of file paths
+    """
+    directory = os.path.dirname(manifest_path)
+    manifest_file = os.path.basename(manifest_path)
+    return [
+        os.path.join(directory, f)
+        for f in os.listdir(directory)
+        if os.path.isfile(os.path.join(directory, f)) and f != manifest_file
+    ]
+
+
+def is_md5_checksum_valid(manifest_file_path: str, processes: int = 1) -> bool:
```

Comment:
absolutely not one for doing in this PR - but one to park in the backlog - possibly supporting all simple checksums that have a tool matching the md5sums format. So I believe my mac comes standard with a `sha1sum` etc.. it might be easy to extend this as basically a table of known easy hashes in the python has library and support a bunch of them alongside bagit

Reply:
Yeah, the BagIt specification supports several algorithms by default, and the BagIt package already provides validation for these as long as the filename is properly formatted with the appropriate algorithm postfix (as per spec). Unfortunately, it seems you can't easily extract just the validation functionality from the BagIt package without also including the full BagIt structure check.

The MD5 implementation here is just a workaround to avoid enforcing the full BagIt structure.

Bagit supported algorithm:
MD5: md5
SHA-1: sha1
sha-256: sha256
sha-512: sha512